### PR TITLE
HOTFIX: don't close or wipe out someone else's state

### PR DESCRIPTION
**Pull request by ableegoldman**
When it comes to actually closing a task we now treat all states exactly the same, and call `StateManagerUtil#closeStateManager` regardless of whether it's in `CREATED` or `RESTORING` or `RUNNING`

Unfortunately `StateManagerUtil` doesn't actually check to make sure that we actually own the lock for this task's state. During a dirty close with eos enabled, we wipe the state -- but in some cases, this means deleting the state out from under another StreamThread who is still in the process of revoking this task

### CODE REVIEW

**Comment by mjsax** [comment]
Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java [hunk 1]
```diff
@@ -100,41 +100,44 @@ static void closeStateManager(final Logger log,
...
+                    try {
+                        stateDirectory.unlock(id);
+                    } catch (final IOException e) {
+                        if (exception == null) {
+                            exception = new ProcessorStateException(String.format("%sFailed to release state dir lock", logPrefix), e);
+                        }
+                    }
                 }
             }
+        } catch (final IOException e) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:125]: Not clear to me why we need a nested try-catch IOException here?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 2]
```diff
@@ -459,7 +459,6 @@ void handleLostAll() {
         final Iterator<Task> iterator = tasks.values().iterator();
         while (iterator.hasNext()) {
             final Task task = iterator.next();
-            final Set<TopicPartition> inputPartitions = task.inputPartitions();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:462]: Is this change intentional?

**Review by guozhangwang — COMMENTED** [review]
While reviewing the PR I had a meta question: the issue would happen if the previous owner thread has not released the lock while the new owner thread is going to close the task. The new thread will try to lock the dir while it is transiting to RESTORING, but the `stateDirectory.lock(id)` will return false immediately if the lock is at hands of another thread within this JVM, which is totally possible if one thread is closing the task while the other is grabbing the task.

So that means in `StateManagerUtil#registerStateStores` it is very possible that the `lock(id)` function would return false, which would be interpreted as fatal error -- more specifically, with incremental protocol since we always close the task before it is reassigned, it may not happen, but with the eager protocol it is very possible this will happen. Right? Is that an overlooked issue?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:462]: Yep, sorry I meant to leave a comment. I noticed this is redundant since `cleanupTask` (invoked by `closeTaskDirty`) does the same thing

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:125]: The only reason would be to distinguish between an `IOException` on `lock`, and one on `unlock`. But I guess that's not actually useful to differentiate, I'll take it out

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:125]: (or during wiping of the state stores, but presumably all this can be discovered through the state trace anyway)

**Comment by ableegoldman** [comment]
>  in StateManagerUtil#registerStateStores it is very possible that the lock(id) function would return false, which would be interpreted as fatal error 

@guozhangwang Can you clarify? If the task fails to lock the directory in `registerStateStores` it throws a `LockException`, which is ultimately caught and handled by the TaskManager in `tryToCompleteRestoration`. It'll just stay in CREATED and try again on the next loop right?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 3]
```diff
@@ -249,32 +247,35 @@ public void testCloseStateManagerOnlyThrowsFirstExceptionWhenClean() throws IOEx
     }
 
     @Test
-    public void testCloseStateManagerDirtyShallSwallowException() throws IOException {
-        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
-
+    public void testCloseStateManagerThrowsExceptionWhenDirty() throws IOException {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java:250]: We actually were already throwing some `ProcessorStateExceptions` up through  close even when unclean, which I think was the cause of a bug we resolved a few weeks ago. Now we just make no assumptions about whether this will throw or not, and catch any exceptions in `StreamTask` / `StandbyTask` if it's a dirty close.

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this

**Comment by guozhangwang** [comment]
test this

**Comment by guozhangwang** [comment]
test this!

**Comment by guozhangwang** [comment]
test this

**Comment by guozhangwang** [comment]
test this

**Comment by guozhangwang** [comment]
> @guozhangwang Can you clarify? If the task fails to lock the directory in `registerStateStores` it throws a `LockException`, which is ultimately caught and handled by the TaskManager in `tryToCompleteRestoration`. It'll just stay in CREATED and try again on the next loop right?

Ah yes, that's right.

**Comment by ableegoldman** [comment]
Failures unrelated:
`ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`
`SuppressionDurabilityIntegrationTest.shouldRecoverBufferAfterShutdown[exactly_once]`

**Comment by guozhangwang** [comment]
test this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java [hunk 4]
```diff
@@ -92,49 +93,46 @@ static void registerStateStores(final Logger log,
...
         }
 
+        final ProcessorStateException exception = firstException.get();
         if (exception != null) {
-            if (closeClean) {
-                throw exception;
-            } else {
-                log.warn("Closing {} task {} uncleanly and swallows an exception", taskType, id, exception);
-            }
+            throw exception;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:135]: Nice clean, now we catch the exception in the caller.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:462]: Thanks! SG.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 5]
```diff
@@ -319,9 +322,47 @@ public void testCloseStateManagerWithStateStoreWipeOutRethrowWrappedIOException(
...
     }
+
+    @Test
+    public void shouldNotStateManagerIfUnableToLockTaskDirectory() throws IOException {
+        expect(stateManager.taskId()).andReturn(taskId);
+
+        expect(stateDirectory.lock(taskId)).andReturn(false);
+
+        stateManager.close();
+        expectLastCall().andThrow(new StreamsException("Should not be trying to close state you don't own!"));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: nit: we can throw AssertionError here to indicate this should not happen.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 6]
```diff
@@ -319,9 +322,47 @@ public void testCloseStateManagerWithStateStoreWipeOutRethrowWrappedIOException(
...
                 "logPrefix:", false, true, stateManager, stateDirectory, TaskType.ACTIVE));
 
-        assertEquals("Failed to wiping state stores for task 0_0", thrown.getMessage());
         assertEquals(IOException.class, thrown.getCause().getClass());
 
         ctrl.verify();
     }
+
+    @Test
+    public void shouldNotStateManagerIfUnableToLockTaskDirectory() throws IOException {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: nit: `shouldNotClose...`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 7]
```diff
@@ -319,9 +322,47 @@ public void testCloseStateManagerWithStateStoreWipeOutRethrowWrappedIOException(
...
+        stateManager.close();
+        expectLastCall().andThrow(new StreamsException("Should not be trying to close state you don't own!"));
+
+        ctrl.checkOrder(true);
+        ctrl.replay();
+
+        replayAll();
+
+        StateManagerUtil.closeStateManager(
+            logger, "logPrefix:", false, true, stateManager, stateDirectory, TaskType.ACTIVE);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: nit: we can set `eosEnabled` to false since we set it to true in the next test?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 8]
```diff
@@ -319,9 +322,47 @@ public void testCloseStateManagerWithStateStoreWipeOutRethrowWrappedIOException(
...
+    public void shouldNotWipeStateStoresIfUnableToLockTaskDirectory() throws IOException {
+        final File unknownFile = new File("/unknown/path");
+        expect(stateManager.taskId()).andReturn(taskId);
+
+        expect(stateDirectory.lock(taskId)).andReturn(false);
+
+        expect(stateManager.baseDir()).andReturn(unknownFile);
+
+        Utils.delete(unknownFile);
+        expectLastCall().andThrow(new StreamsException("Should not be trying to wipe state you don't own!"));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: Ditto here, we can use AssertionError

**Review by guozhangwang — COMMENTED** [review]
Made another pass, after addressing this should be good to merge.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: good call

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this please
